src/segwit.rs

```diff
@@ -0,0 +1,227 @@
+/// let (_hrp, _witness_version, _witness_program) = segwit::decode(address).expect("failed to decode address");
+/// ```
+#[cfg(feature = "alloc")]
+pub fn decode(s: &str) -> Result<(Hrp, Fe32, Vec<u8>), SegwitHrpstringError> {
+    let segwit = SegwitHrpstring::new(s)?;
+    Ok((segwit.hrp(), segwit.witness_version(), segwit.byte_iter().collect::<Vec<u8>>()))
+}
+
+/// Encodes a segwit address.
+///
+/// Does validity checks on the `witness_version` and length checks on the `witness_program`.
+///
+/// As specified by [`BIP-350`] we use the [`Bech32m`] checksum algorithm for witness versions 1 and
+/// above, and for witness version 0 we use the original ([`BIP-173`]) [`Bech32`] checksum
+/// algorithm.
+///
+/// See also [`encode_v0`] or [`encode_v1`].
+///
+/// [`Bech32`]: crate::primitives::Bech32
+/// [`Bech32m`]: crate::primitives::Bech32m
+/// [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+/// [BIP-350]: <https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki>
+#[cfg(feature = "alloc")]
+pub fn encode(
+    hrp: &Hrp,
+    witness_version: Fe32,
+    witness_program: &[u8],
+) -> Result<String, EncodeError> {
+    segwit::validate_witness_version(witness_version)?;
+    segwit::validate_witness_program_length(witness_program.len(), witness_version)?;
+
+    let mut buf = String::new();
+    encode_to_fmt_unchecked(&mut buf, hrp, witness_version, witness_program)?;
+    Ok(buf)
+}
+
+/// Encodes a segwit version 0 address.
+#[cfg(feature = "alloc")]
+pub fn encode_v0(hrp: &Hrp, witness_program: &[u8]) -> Result<String, EncodeError> {
+    encode(hrp, Fe32::Q, witness_program)
+}
+
+/// Encodes a segwit version 1 address.
+#[cfg(feature = "alloc")]
+pub fn encode_v1(hrp: &Hrp, witness_program: &[u8]) -> Result<String, EncodeError> {
+    encode(hrp, Fe32::P, witness_program)
+}
+
+/// Encodes a segwit address to a writer ([`fmt::Write`]).
+///
+/// Does not check the validity of the witness version and witness program lengths (see
+/// the [`crate::primitives::segwit`] module for validation functions).
+pub fn encode_to_fmt_unchecked(
+    fmt: &mut dyn fmt::Write,
+    hrp: &Hrp,
+    witness_version: Fe32,
+    witness_program: &[u8],
+) -> fmt::Result {
+    match witness_version {
+        Fe32::Q => {
+            for c in witness_program
+                .iter()
+                .copied()
+                .bytes_to_fes()
+                .with_checksum::<Bech32>(hrp)
+                .with_witness_version(Fe32::Q)
+                .chars()
+            {
+                fmt.write_char(c)?;
+            }
+        }
+        version => {
+            for c in witness_program
+                .iter()
+                .copied()
+                .bytes_to_fes()
+                .with_checksum::<Bech32m>(hrp)
+                .with_witness_version(version)
+                .chars()
+            {
+                fmt.write_char(c)?;
+            }
```

Comment:
Cool, I was unsure about the capitialisation thing because one of the bips (173?) specifically says that encoders should encode as lowercase and leave capitialisation to higher layers. It does fit in better with Rust to have it done here, like you say, especially for the no-alloc functions. I'll have another go and see what I come up with.

Reply:
Yeah, that's a weird comment for the BIP to make. Maybe it shoud look at the log in its own eye before lecturing us about layer separation.